src/drivers/net/dwc_gmac/dwc_gmac.c

```diff
@@ -0,0 +1,596 @@
+}
+
+static int dwc_mdio_write(struct net_device *dev, uint8_t regAddr, uint16_t data) {
+	struct dwc_priv *dwc_priv;
+	uint32_t reg;
+	uint32_t phy;
+	uint16_t mii;
+	size_t cnt;
+
+	assert(dev);
+
+	dwc_priv = dev->priv;
+
+	reg = regAddr;
+	phy = dwc_priv->phy_id;
+
+	mii = ((phy & DWC_GMAC_GMII_ADDRESS_PA_MASK) << DWC_GMAC_GMII_ADDRESS_PA_SHIFT)
+	    | ((reg & DWC_GMAC_GMII_ADDRESS_GR_MASK) << DWC_GMAC_GMII_ADDRESS_GR_SHIFT)
+	    | (dwc_priv->mii_clk << DWC_GMAC_GMII_ADDRESS_CR_SHIFT)
+	    | DWC_GMAC_GMII_ADDRESS_GB | DWC_GMAC_GMII_ADDRESS_GW;
+
+	dwc_reg_write(dwc_priv, DWC_GMAC_MII_DATA, data);
+	dwc_reg_write(dwc_priv, DWC_GMAC_MII_ADDR, mii);
+
+	for (cnt = 0; cnt < 1000; cnt++) {
+		if (!(dwc_reg_read(dwc_priv, DWC_GMAC_MII_ADDR) & DWC_GMAC_GMII_ADDRESS_GB)) {
+			break;
+                }
+		delay(10);
+	}
+
+	return 0;
+}
+
+static void dwc_set_phyid(struct net_device *dev, uint8_t phyid) {
+	struct dwc_priv *dwc_priv;
+
+	assert(dev);
+
+	dwc_priv = dev->priv;
+	dwc_priv->phy_id = phyid;
+}
+
+static int dwc_set_speed(struct net_device *dev, int speed) {
+	return 0;
+}
+
+static const struct net_driver dwc_drv_ops = {
+	.xmit        = dwc_xmit,
+	.start       = dwc_open,
+	.set_macaddr = dwc_set_macaddr,
+	.mdio_read   = dwc_mdio_read,
+	.mdio_write  = dwc_mdio_write,
+	.set_phyid   = dwc_set_phyid,
+	.set_speed   = dwc_set_speed
+};
+
+
+static uint32_t dwc_setup_rxdesc(int idx) {
+	struct dma_extended_desc *desc;
+	int next_desc;
+
+	assert(idx < RX_DESC_QUANTITY);
+
+	desc = &rx_descs_pool[idx];
+
+	desc->basic.des2 = (uint32_t)&rx_buffers[idx][0];
+
+	if (idx < (RX_DESC_QUANTITY - 1)) {
+		next_desc = idx + 1;
+		desc->basic.des1 = ERDES1_SECOND_ADDRESS_CHAINED | ETH_FRAME_LEN;
+	} else {
+		next_desc = 0;
+		desc->basic.des1 = ERDES1_SECOND_ADDRESS_CHAINED | ETH_FRAME_LEN /*| ERDES1_END_RING */;
+	}
+
+	desc->basic.des3 = (uint32_t)&rx_descs_pool[next_desc];
+
+	dcache_flush(desc, sizeof(*desc));
+	desc->basic.des0 = RDES0_OWN;
+	dcache_flush(desc, sizeof(*desc));
+
+	return next_desc;
+}
+
+
+static inline int dwc_rxfinish_locked(struct net_device *dev_id) {
+	int cur_desc;
+	struct sk_buff *skb;
+	struct dwc_priv *priv;
+	struct dma_extended_desc *desc;
+	int len;
+
+	priv = dev_id->priv;
+
+	for(;;) {
+		cur_desc = priv->rxdesc_id;
+
+
+		desc = &((struct dma_extended_desc *)priv->rxdesc_ring_paddr)[cur_desc];
+		dcache_inval(desc, sizeof(*desc));
+
+		if (desc->basic.des0 & RDES0_OWN) {
+			return 0;
+		}
+
+		len = (desc->basic.des0 >> 16) & 0x1FFF;
+		dcache_inval((void *)desc->basic.des2, len);
+
+
+		skb = skb_alloc(len);
+		if (!skb) {
+			log_error("couldn't allocate skb");
+			return -ENOMEM;
+		}
+		memcpy(skb_data_cast_in(skb->data), (void *)desc->basic.des2, len);
+
+		skb->len = len - 4; /* without CRC */
+		skb->dev = dev_id;
+
+		show_packet(skb_data_cast_in(skb->data), skb->len, "rx");
+
+		netif_rx(skb);
+
+		dwc_setup_rxdesc(priv->rxdesc_id);
+
+		if (priv->rxdesc_id < (RX_DESC_QUANTITY - 1)) {
```

Comment:
Overflow may occur here (rxdesc_id may be equal `RX_DESC_QUANTITY`)

I recomment to use `priv->rxdesc_id %= RX_DESC_QUANTITY;`